### PR TITLE
👷 Refactor ➾ Implement Code Generation for Zod Types in Taqueria Protocol

### PR DESCRIPTION
**Pull request by mweichert**
# 🌮 Taqueria PR

## 🪁 Description

The `@taqueria/protocol` package is used to contain all of our custom types shared between the CLI and SDK.

Before this PR, for each type, we would define a schema in Zod and then have a type inferred from the schema. This PR reverses that - we define a type, with some Zod specific annotations, and then use code generation tools to generate our zod schemas and type modules.

## 🪂 Pre-Merge Checklist (Definition of Done)

🚦 Required to merge:

- [x] ⛱️ I have completed this PR template in full and updated the title appropriately
- [x] ⛵ My code builds cleanly, and I have manually tested the changes
- [ ] 🏄‍♂️ Another team member has built this branch and done manual testing on the change
- [ ] 🏖️ New and existing unit tests pass locally and in CI
- [ ] 🔱 The test plan has been implemented and verified by an SDET
- [x] 🦀 Automated tests have been written and added to this PR
- [x] 🐬 I have commented my code, particularly in hard-to-understand areas
- [x] 🤿 Corresponding changes have been made to all documentation
- [x] 🐚 Required changes to the VScE have been made
- [ ] 🪸 Required updates to scaffolds have been made
- [ ] 🚢 The release checklist has been completed

## 🛩️ Summary of Changes

Rick introduced a new package called `@taqueria/protocol-types` was introduced. I think this was a means of segregating the refactor work from the original implementation.  The `@taqueria/protocol-types/types.ts` file contains our types which serves as the source of truth / input to our code generators. You can invoke the code generator by running, `npm run build-types` (from the root of the taqueria repo), which will create our schemas in `@taqueria/protocol-types/types-zod.ts`, and our strict type modules in `@taqueria/protocol-types/out/types/`.

I took over Rick's work to do the following:
- Adjust the `@taqueria/protocol` and `@taqueria/node-sdk` packages to use the types from `@taqueria/protocol-types`
- Adjust anything to conform with the new types. I tried to keep this to a minimum to reduce the overall amount of change.
- Some type modules in @taqueria/protocol`, such as EphermeralState.ts, aren't using generated types from the `@taqueria/protocol-types` package. This is because these type modules include logic that cannot be easily generated. Rick is working on this separately, I think in a different branch.

Thus, this PR serves as the first iteration of 2 or 3 efforts of work, the second iteration being one that improves our code generation in such a way that we can use generated types for all type modules in @taqueria/protocol.

## 🎢 Test Plan

We're just reusing our existing test suites.

## 🛸 Type of Change

- [ ] 🧽 Chore ➾
- [ ] 🛠️ Fix ➾
- [ ] ✨ Feature ➾
- [x] 👷 Refactor ➾
- [ ] 🧪 Pre-Release ➾
- [ ] 🚀 Release ➾

### CODE REVIEW

**Comment by mweichert** [comment]
NOTE - I have 3 failing test suites that I'm working on atm...
taqueria-plugin-taquito-flextesa.spec.ts
taqueria-plugin-jest.spec.ts
taqueria-plugin-taquito.spec.ts

**Review by ricklove — COMMENTED** [review]
All the changes look good.

The e2e and integration tests are not yet passing for me - still running and seeing if I can identify the issue

**Comment by jchenche** [comment]
@mweichert Can you do a merge from main as a lot of the changes are outdated?

**Review by ricklove — CHANGES_REQUESTED** [review]
Looks like the changes around ProxyTaskArgs broke the intergration tests:

- expected string, got number

The integration tests are working in the last commit after fixing integration tests, but not working in last commit

**Comment by ricklove** [comment]
Looks like this is yargs causing the issue:

image

**Comment by ricklove** [comment]
Looks like it needs `string: true`

**Comment by jchenche** [comment]
yargs sneak attack again! 😭 @ricklove

File: taqueria-protocol-types/out/types-strict.ts [hunk 1]
```diff
@@ -188,7 +188,13 @@ type PluginProxyResponseRaw = void | PluginJsonResponse;
 /** @default none */
 export type PluginResponseEncoding = { __type: PluginResponseEncoding } & PluginResponseEncodingRaw;
 type PluginResponseEncodingRaw = 'none' | 'json' | 'application/json';
-export type SanitizedArgs = { __type: SanitizedArgs } & {
+
+/**
+ * @min 100
```
Comment (jchenche) [inline taqueria-protocol-types/out/types-strict.ts:193]: What syntax is this? Any docs on this?

**Review by jchenche — CHANGES_REQUESTED** [review]
This PR is very outdated and needs to merge main in and resolve all conflicts

**Review by jchenche — COMMENTED** [review]
Peeking type definitions is much better. Thanks @mweichert @ricklove 🎉

File: taqueria-protocol-types/out/types/ConfigArtifactsDir.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// Generated file: Do not edit
```
Comment (jchenche) [inline taqueria-protocol-types/out/types/ConfigArtifactsDir.ts:1]: We shouldn't track anything within the `/out` folder right? Or anything generated

Comment (ricklove) [reply taqueria-protocol-types/out/types/ConfigArtifactsDir.ts]: right - this was added to gitignore

Comment (jchenche) [reply taqueria-protocol-types/out/types/ConfigArtifactsDir.ts]: @ricklove And we probably need another commit to untrack it too right?

**Pull request closed without merging**